File: arch-ui/src/wire/session.ts
import { useSyncExternalStore } from "react";
import { ask, boardLine, dropTurn, getChat, hasAsk, nextTurnId, patchTurn, push, queueAsk, resetChat, say, setChat, settleAskInMessage, spendAsk, you } from "../board/chat";
import type { PickerOption } from "../board/picker";
import { splitTask } from "./task";
import { flash } from "../board/ui";
import type { ArchState, ConnState, Frontier, Incoming, ReadyEvent, ScribeEvent } from "./types";

/**
 * Harness truth. Every field here is written by an event and never by the page.
 *
 * `arch_state` replaces the whole design wholesale — the page never diffs it.
 * The only thing derived from a push is what to *highlight*: `changed` names
 * the id that just moved.
 */

export interface SessionState {
  conn: ConnState;
  ready: ReadyEvent | null;
  arch: ArchState | null;
  noticing: string[];
  changed: { kind: string; id: string } | null;
  running: boolean;
  handedOff: boolean;
  /** how much the user has drawn that the architect has not been shown */
  pendingEdits: number;
  /** everything that was already on the board when this page connected —
   *  boxes by id, wires as "src>dst". A refresh delivers the whole design at
   *  once, and animating it would claim eleven boxes were just created. These
   *  are born already-in. */
  bornWith: Record<string, true>;
  totalIn: number;
  totalOut: number;
  /** what is still askable, and what the user closed — from the harness */
  frontier: Frontier | null;
  /** the scribe's state, when one is drawing the board a step behind */
  scribe: ScribeEvent | null;
  /** a scribe is on: the board's lines belong to its strip, not the thread */
  scribeOn: boolean;
  /** the research turn's steps, in order */
  research: { step: string; text: string; done: boolean }[];
}

let state: SessionState = {
  conn: "connecting",
  ready: null,
  arch: null,
  noticing: [],
  changed: null,
  running: false,
  handedOff: false,
  pendingEdits: 0,
  bornWith: {},
  totalIn: 0,
  totalOut: 0,
  frontier: null,
  scribe: null,
  scribeOn: false,
  research: [],
};

const listeners = new Set<() => void>();
const emit = () => { for (const l of listeners) l(); };
export const getSession = () => state;
const set = (patch: Partial<SessionState>) => { state = { ...state, ...patch }; emit(); };

export function useSession(): SessionState {
  return useSyncExternalStore(
    (cb) => { listeners.add(cb); return () => listeners.delete(cb); },
    getSession,
    getSession,
  );
}

/* ── turning harness events into a conversation ───────────────────────── */

/** The one line the architect's machinery gets, and the boxes it touched so
 *  "show me" has somewhere to go. */
function toolLine(name: string, details: Record<string, unknown> | null | undefined) {
  const d = (details ?? {}) as { subjects?: unknown; nodes?: unknown; summary?: unknown };
  const subjects = Array.isArray(d.subjects) ? d.subjects.map(String) : [];
  const nodes = Array.isArray(d.nodes) ? d.nodes.map(String) : [];
  /* the tool's own first line, which already says what it did; the tool name
     alone tells you a call happened but not what changed */
  const said = typeof d.summary === "string" && d.summary ? d.summary : name;
  return { text: "board · " + said.replace(/^Board: /, ""), ids: subjects.length ? subjects : nodes };
}

/** The turn the architect is speaking into, so its tool lines land under the
 *  text that describes them rather than in a block of their own. */
let openTurnId: number | null = null;
let thinkingId: number | null = null;
/** the research turn's list, while it is the live indicator */
let progressId: number | null = null;

/** Text arrives a token at a time. It lands in a turn of its own that grows,
 *  rather than appearing whole when the turn is already over — the waiting is
 *  most of the turn, and a page that shows nothing during it reads as stuck. */
let streamId: number | null = null;
let streamText = "";

function finishStream(final?: string) {
  if (streamId === null) return null;
  const id = streamId;
  const text = final ?? streamText;
  streamId = null;
  streamText = "";
  if (!text) { dropTurn(id); return null; }
  patchTurn(id, { lines: [text] } as never);
  return id;
}

function stopThinking() {
  if (thinkingId !== null) { dropTurn(thinkingId); thinkingId = null; }
}

function startThinking() {
  if (thinkingId === null) thinkingId = push({ t: "thinking", id: nextTurnId() });
}

export function applyEvent(ev: Incoming): void {
  switch (ev.type) {
    case "ready":
      set({
        ready: ev,
        conn: state.conn === "complete" ? "complete" : "connected",
        totalIn: ev.input_tokens ?? 0,
        totalOut: ev.output_tokens ?? 0,
        scribeOn: Boolean(ev.scribe),
      });
      break;

    case "scribe":
      set({ scribe: ev });
      break;

    case "arch_state": {
      const handedOff = ev.status === "handed_off";
      /* Only a replayed push is history. A live one is something that just
         happened, however early in the session it happens to be. */
      const bornWith = ev.replayed
        ? Object.fromEntries([
            ...Object.keys(ev.state.nodes).map((id) => [id, true as const]),
            ...ev.state.edges.map((e) => [`${e.src}>${e.dst}`, true as const]),
          ])
        : state.bornWith;
      set({
        arch: ev.state,
        noticing: ev.noticing ?? [],
        changed: ev.changed,
        pendingEdits: ev.pending_edits ?? 0,
        frontier: ev.frontier ?? state.frontier,
        bornWith,
        ...(handedOff && !state.handedOff ? { handedOff: true, conn: "complete" as ConnState } : {}),
      });
      /* The one question on the table, if there is one. The harness re-pushes
         its whole state on every change, so the id is what makes this idempotent
         — and because it only ever sends the earliest open question, the next
         one appears here exactly when the last is answered. */
      if (ev.ask && !hasAsk(ev.ask.id)) {
        if (openTurnId === null) openTurnId = push({ t: "say", id: nextTurnId(), lines: [] });
        ask(openTurnId, ev.ask);
      }
      break;
    }

    case "harness_event": {
      const { event, data } = ev;
      if (event === "run_start") {
        if (data.task) {
          /* A message can be drawn, typed, pointed, or all three — the harness
             sends them as one turn, so the page has to show every half or the
             typed words disappear from the record. */
          const { drew, about, picked, typed } = splitTask(data.task);
          /* A bare pick says nothing here: the row it answered has already
             collapsed into the answered block above, which is the record. A
             turn reading "you · picked" under it is the same fact twice. */
          const bare = picked.length && !typed && !drew.length && !about.length;
          if (bare) {
            /* nothing to add to the thread */
          } else if (drew.length || about.length || picked.length) {
            const via = !typed && drew.length && !picked.length ? "on the board"
              : !typed && picked.length ? "picked" : undefined;
            you(typed || undefined, undefined, via, drew, about);
          } else {
            you(data.task);
          }
          /* A pick settles the question it answered; a typed reply with no
             pick means they answered in prose — the rows stay as the record
             of what was offered, but none of them was taken. */
          for (const t of getChat().turns) {
            if (t.t === "say" && t.ask && !t.ask.spent) {
              if (picked.length) spendAsk(t.id, picked[0]);
              else settleAskInMessage(t.id);
            }
          }
        }
        openTurnId = null;
        finishStream();
        set({ running: true });
        startThinking();
      } else if (event === "assistant_delta") {
        stopThinking();
        streamText += data.text || "";
        if (streamId === null) streamId = push({ t: "say", id: nextTurnId(), lines: [streamText] });
        else patchTurn(streamId, { lines: [streamText] } as never);
      } else if (event === "assistant") {
        stopThinking();
        /* the streamed copy and the final one are the same words — settle the
           turn that has been growing rather than saying it all twice */
        const settled = finishStream(data.content || undefined);
        if (settled !== null) openTurnId = settled;
        else if (data.content) openTurnId = say([data.content]);
        else if (openTurnId === null) openTurnId = push({ t: "say", id: nextTurnId(), lines: [] });
        if (data.tool_calls?.length) startThinking();
      } else if (event === "tool_result") {
        const line = toolLine(String(data.name ?? "tool"), data.details);
        /* With a scribe on, the board's lines belong to its strip: the thread
           carries the argument only. The halo still says what moved. */
        if (!state.scribeOn) {
          if (openTurnId === null) openTurnId = push({ t: "say", id: nextTurnId(), lines: [] });
          boardLine(openTurnId, line.text, line.ids);
        }
        /* halo everything that call touched, so a change that lands while you
           are reading elsewhere is still visible when you look back */
        if (line.ids.length) flash(line.ids);
      } else if (event === "research") {
        /* the research turn: one line per kind of work, in place of the
           thinking dots, kept afterwards as the record of the first minute */
        const step = String(data.step ?? "");
        const text = String(data.text ?? step);
        const done = Boolean(data.done);
        const steps = state.research.some((r) => r.step === step)
          ? state.research.map((r) => (r.step === step ? { ...r, text, done } : r))
          : [...state.research, { step, text, done }];
        set({ research: steps });
        stopThinking();
        if (progressId === null) progressId = push({ t: "progress", id: nextTurnId(), steps });
        else patchTurn(progressId, { steps } as never);
      } else if (event === "abort") {
        stopThinking();
        say([`_The turn stopped: ${data.reason || "no reason given"}._`]);
      }
      break;
    }

    case "turn_end":
      stopThinking();
      finishStream();
      openTurnId = null;
      if (progressId !== null) {
        /* the list stays as the record; nothing in it is live any more */
        const steps = state.research.map((r) => ({ ...r, done: true }));
        patchTurn(progressId, { steps } as never);
        set({ research: steps });
        progressId = null;
      }
      if (ev.status === "interrupted") say([stopCopy(ev.reason)]);
      set({
        running: false,
        totalIn: ev.input_tokens ?? state.totalIn,
        totalOut: ev.output_tokens ?? state.totalOut,
      });
      break;

    case "error":
      stopThinking();
      finishStream();
      say([`_${ev.message || "the turn failed"}_`]);
      set({ running: false });
      break;

    case "bye":
      over = true;
      source?.close();
      source = null;
      if (!state.handedOff) set({ conn: "disconnected", running: false });
      break;
  }
}

/** Why a turn stopped, in the page's words. "you interrupted it" is said
 *  only when the harness says it was the user. */
export function stopCopy(reason: unknown): string {
  if (reason === "user") return "_You stopped the turn._";
  if (reason === "shutdown") return "_The session is closing._";
  return "_The turn stopped._";
}

/* ── talking back ─────────────────────────────────────────────────────── */

async function post(path: string, body: unknown): Promise<Response | null> {
  try {
    return await fetch(path, { method: "POST", body: JSON.stringify(body ?? {}) });
  } catch {
    return null; // surfaced through the connection state, not a toast
  }
}

/** `subjects` is what was selected on the board when Send was pressed. The
 *  harness turns the ids into the boxes' own details and puts them in front of
 *  the message, so a question that points at something arrives knowing what it
 *  pointed at. */
export function sendInput(text: string, subjects: string[] = []): void {
  void post("/input", subjects.length ? { text, subjects } : { text });
}

/**
 * Answering with a row of the picker.
 *
 * The answer goes to the harness, not to the message box: the harness settles
 * the question on its own state and starts the turn itself, prefixed so the
 * transcript records "picked" rather than words the user never typed. That
 * order matters — the state is what decides which question is next, so the
 * page can never get ahead of it.
 */
export function sendAnswer(host: number, id: string, option: PickerOption): void {
  /* While the architect is still writing, the harness parks the answer until
     the turn ends. The dock says so; the turn the answer starts is what
     collapses the question (run_start carries the pick). */
  if (state.running) queueAsk(host, option.label, option.value);
  else spendAsk(host, option.label, option.value);  // it reads as done the frame it happens
  void post("/answer", { id, value: option.value });
}

/** The user closes a branch from the frontier, or reopens one. */
export function closeBranch(op: "settle" | "out_of_scope" | "reopen", id: string): Promise<string | null> {
  return mutate({ op, id });
}

/**
 * Send what the user drew.
 *
 * Explicit on purpose. An earlier version watched for the board to go quiet
 * and submitted by itself, which spent a model call every time the user paused
 * to think and asked the architect to respond to half-finished gestures. Only
 * the person drawing knows when they have finished a sentence.
 */
export function sendBoard(): void {
  void post("/board", {});
}

export function interrupt(): void {
  void post("/interrupt", {});
}

/**
 * A user edit: sent, and rolled back onto the page if the harness refuses.
 *
 * There is no optimistic copy of the design here — `arch_state` is a full
 * replacement and the push that a successful mutation triggers arrives in
 * milliseconds. What the page *does* hold optimistically is position, because
 * a box that snaps back under your cursor while the round trip lands is worse
 * than one that arrives a frame late.
 */
export async function mutate(payload: Record<string, unknown>): Promise<string | null> {
  const res = await post("/mutate", payload);
  if (res === null) return "could not reach the harness — the edit was not applied";
  if (res.ok) return null;
  const body = await res.json().catch(() => ({} as { error?: string }));
  return String(body.error || `the harness refused the edit (${res.status})`);
}

/** A refusal belongs in the conversation: it is the harness disagreeing, and
 *  that is the same channel everything else disagrees on. */
export function refusal(message: string): void {
  say([`_${message}_`]);
}

/* ── the connection ───────────────────────────────────────────────────── */

/**
 * One SSE connection at a time, for the life of the page. The harness replays
 * a joiner (ready, a bounded transcript buffer, the latest arch_state), so a
 * refresh mid-session rebuilds everything without special-casing — and so
 * does a reconnect: a dropped connection used to be the end of the page, its
 * composer announcing the harness gone while the server sat there serving.
 * Now it is retried with backoff, the transcript cleared just before the
 * replay so nothing lands twice, and only a `bye` is final.
 */
const BACKOFF_MAX_MS = 15000;
let source: EventSource | null = null;
let retryTimer: ReturnType<typeof setTimeout> | null = null;
let attempt = 0;
/** the harness said goodbye: nothing to reconnect to */
let over = false;

function resetForReplay(): void {
  resetChat();
  openTurnId = null; thinkingId = null; streamId = null; streamText = ""; progressId = null;
  set({ bornWith: {}, research: [] });
}

export function connect(): void {
  if (retryTimer) { clearTimeout(retryTimer); retryTimer = null; }
  if (typeof EventSource === "undefined") return;
  source?.close();
  const es = new EventSource("/events");
  source = es;
  es.onopen = () => {
    attempt = 0;
    if (state.conn === "reconnecting") resetForReplay();
  };
  es.onmessage = (e) => {
    let parsed: Incoming;
    try {
      parsed = JSON.parse(e.data) as Incoming;
    } catch {
      return; // a malformed frame must never take the page down
    }
    applyEvent(parsed);
  };
  es.onerror = () => {
    if (source !== es) return; // an older connection, already replaced
    es.close();
    source = null;
    if (over || state.handedOff || state.conn === "complete") {
      if (state.conn !== "complete") set({ conn: "disconnected", running: false });
      return;
    }
    const delay = Math.min(BACKOFF_MAX_MS, 1000 * 2 ** attempt);
    attempt++;
    if (state.conn !== "reconnecting") set({ conn: "reconnecting" });
    retryTimer = setTimeout(connect, delay);
  };
}

/** How the page is doing on the wire, for tests. */
export const connection = () => ({ attempt, retrying: retryTimer !== null, over });

/** Tests: back to a fresh page. */
export function resetSession(): void {
  if (retryTimer) { clearTimeout(retryTimer); retryTimer = null; }
  source?.close(); source = null; attempt = 0; over = false;
  state = { ...state, conn: "connecting", arch: null, noticing: [], changed: null, running: false, handedOff: false,
    pendingEdits: 0, bornWith: {}, frontier: null, scribe: null, scribeOn: false, research: [] };
  openTurnId = null; thinkingId = null; streamId = null; streamText = ""; progressId = null;
  emit();
}

export const chatIsOpen = () => getChat().open;
export const markChat = setChat;

File: src/bird/harnesses/arch/session.py
"""ArchSession — the harness-owned wrapper the arch tools mutate through.

Holds the ArchState plus the little wiring the tools need: the arch_state event
sink (straight to the transport — arch_state is a top-level protocol event, not
a harness_event) and per-mutation persistence.

There is no broker here, and no critic. There are no gates to broker: the
session ends when the user says it is done, so nothing blocks a turn waiting on
a modal. And the background critic is gone by design — the user is in the room
the whole time, which makes them the critic; the deterministic checks that used
to feed it live in `derive.py` and reach the architect as material for its next
recommendation instead of as a third voice.

The write lock stays: `apply_mutation` arrives on an HTTP thread, so a page edit
must not interleave with a tool call into a half-written state file.
"""

from __future__ import annotations

import json
import threading
from pathlib import Path
from collections.abc import Sequence
from typing import Any, Callable

from ..picker import Ask
from . import derive, render
from .state import ArchState

STATE_FILENAME = "arch_state.json"

# How many turns a user's edit stays in front of the architect.
#
# The tracker is re-rendered every loop iteration, so draining it on the first
# render would show an edit once and then delete it from the conversation
# mid-turn — the architect would glimpse it and lose it. Holding each edit for
# a few renders means it is genuinely read, without lingering forever and
# inviting the architect to react to the same edit twice.
SHOW_EDITS_FOR = 3

# How a board edit announces itself as a turn. The page reads this to render it
# as something the user *did* rather than something they typed.
BOARD_EDIT_PREFIX = "[the user changed the board]"

# How a message says which boxes the user had selected when they sent it. The
# page reads this back so the selection shows as context on the turn rather
# than as words the user typed.
FOCUS_PREFIX = "[the user is pointing at]"

# A page can only select one box today. The cap is here because the id list
# arrives over HTTP and nothing else bounds it.
MAX_FOCUS = 8

# How a picked row announces itself as a turn. The page reads this back so the
# transcript records "picked" rather than words the user never typed — the
# other half of this contract is PICK_PREFIX in arch-ui/src/wire/task.ts.
PICK_PREFIX = "[the user picked]"


class ArchSession:
    def __init__(
        self,
        state: ArchState | None = None,
        run_dir: Path | None = None,
        on_state: Callable[[dict[str, Any]], None] | None = None,
    ) -> None:
        self.state = state or ArchState()
        self.run_dir = run_dir
        self.on_state = on_state
        self._write_lock = threading.RLock()  # touched() nests inside apply_mutation
        self._user_edits: list[list[Any]] = []  # [text, renders_so_far]

    def state_event(self, changed: dict[str, str] | None = None) -> dict[str, Any]:
        """The full-replacement push. `status` is what the transport's stop
        condition reads — a session is open until it is handed off."""
        ask = self.pending_ask()
        return {
            "type": "arch_state",
            "status": "handed_off" if self.state.handed_off else "open",
            "state": self.state.to_dict(),
            "renders": render.render_all(self.state),
            # what the harness noticed, keyed for the page to badge a node. The
            # architect gets the same material through its own state note; the
            # page shows it so the user can point at it, never as a task list.
            "noticing": derive.coverage(self.state),
            # how much the user has drawn that the architect has not been shown
            # yet. The page badges its submit with this, and reading it off the
            # harness means a refresh does not lose the count.
            "pending_edits": self.pending_edit_count(),
            # what is still askable, the fork if open, and what the user closed
            # — the same walk the architect's note is built from, so the user
            # sees why it asks what it asks and can end a branch on purpose
            "frontier": derive.frontier_struct(self.state),
            # The one question on the table, as the picker payload both this
            # page and the design workbench render. Deliberately the earliest
            # open one and no other: two pickers in a thread is two questions
            # asked at once, which is how a user ends up answering neither.
            "ask": ask.payload() if ask is not None else None,
            "changed": changed,
        }

    def touched(self, changed_kind: str | None = None, changed_id: str | None = None) -> None:
        """Persist + push the full state after every mutation (mid-turn)."""
        with self._write_lock:
            if self.run_dir is not None:
                self.run_dir.mkdir(parents=True, exist_ok=True)
                (self.run_dir / STATE_FILENAME).write_text(
                    json.dumps(self.state.to_dict(), ensure_ascii=False, indent=1),
                    encoding="utf-8",
                )
            if self.on_state is not None:
                changed = None
                if changed_kind is not None and changed_id is not None:
                    changed = {"kind": changed_kind, "id": changed_id}
                self.on_state(self.state_event(changed))

    def apply_mutation(self, payload: dict[str, Any]) -> dict[str, Any]:
        """A user edit made on the page, applied through the model's own code
        paths. Duck-typed on purpose: the pump finds this by name on `ctx.arch`,
        so `POST /mutate` needs no knowledge that the arch harness exists.
        Raises MutationError, which the transport turns into a 400."""
        from .mutate import apply_mutation

        with self._write_lock:  # touched() re-enters it on this thread
            return apply_mutation(self, payload)

    # ---- what the user did to the board ----

    def note_user_edit(self, text: str) -> None:
        """Record something the person changed on the page.

        The design itself already reflects it — the state is shared — but the
        state is a snapshot and cannot say *who* moved something or that a
        label is new. A rename or a pinned note is the user talking; it should
        reach the architect as plainly as a typed message would.
        """
        with self._write_lock:
            for entry in self._user_edits:
                if entry[0] == text:
                    entry[1] = 0  # said again: start its life over
                    return
            self._user_edits.append([text, 0])

    def take_user_edits(self) -> list[str]:
        """The edits to show this turn, ageing out the ones already read."""
        with self._write_lock:
            out = [text for text, _ in self._user_edits]
            for entry in self._user_edits:
                entry[1] += 1
            self._user_edits = [e for e in self._user_edits if e[1] < SHOW_EDITS_FOR]
            return out

    def pending_edit_count(self) -> int:
        """Edits the architect has not seen. Only these are worth submitting."""
        with self._write_lock:
            return sum(1 for _, renders in self._user_edits if renders == 0)

    def compose_activity_prompt(self) -> str | None:
        """A turn's worth of prompt from what the user just drew, or None.

        Only edits the architect has not already seen count. An edit made while
        a turn was running has reached it through the pinned note, and asking it
        to respond to the same gesture a second time is worse than not asking.
        """
        with self._write_lock:
            fresh = [text for text, renders in self._user_edits if renders == 0]
            if not fresh:
                return None
            # they are delivered now — by this prompt rather than by the note
            self._user_edits = [e for e in self._user_edits if e[1] > 0]
        # Just what they did. How to answer it is in the system prompt, where
        # it costs nothing to repeat; restating it every turn is tokens spent
        # on something the architect already knows.
        lines = "\n".join(f"- {text}" for text in fresh)
        return f"{BOARD_EDIT_PREFIX}\n{lines}"

    def describe_subjects(self, ids: Sequence[str]) -> str | None:
        """The boxes the user had selected, described for the architect.

        Selection lives in the page — the harness has no idea what is
        highlighted until a message says so. Sending each box's own details
        rather than just its id is the whole point: it lets "why this one?" be
        answered without a lookup and without guessing which box "this" meant.

        Unknown ids are dropped rather than reported. A selection that no
        longer resolves is the page being a moment behind the graph, which is
        not something to make the architect explain.
        """
        with self._write_lock:
            blocks = [
                self._describe_node(nid)
                for nid in list(dict.fromkeys(ids))[:MAX_FOCUS]
            ]
        kept = [b for b in blocks if b]
        if not kept:
            return None
        return "\n".join([FOCUS_PREFIX, *kept])

    def _describe_node(self, node_id: str) -> str | None:
        """One box: a `- label` bullet the page reads back for the transcript,
        plus indented lines it ignores and the architect does not."""
        node = self.state.nodes.get(node_id)
        if node is None:
            return None
        named = [
            self.state.approaches[a].name
            for a in node.approaches
            if a in self.state.approaches
        ]
        facts = [f"id: {node.id}", f"kind: {node.kind}", f"depth: {node.depth}"]
        facts.append(
            "approach: " + (", ".join(named) if named else "shared by every approach")
        )
        if self.state.is_greyed(node):
            facts.append("on an approach that was not taken")
        if node.existing:
            facts.append("already in the repo — background, not ours to design")

        lines = [f"- {node.label}", "  " + " · ".join(facts)]
        for name, value in (
            ("owns", node.responsibility),
            ("built on", node.tech),
            ("inside", node.detail),
            ("noted", node.notes),
        ):
            if value:
                lines.append(f"  {name}: {value}")

        wires = [
            f"→ {self._node_name(e.dst)}" + (f" ({e.label})" if e.label else "")
            for e in self.state.edges if e.src == node.id
        ] + [
            f"← {self._node_name(e.src)}" + (f" ({e.label})" if e.label else "")
            for e in self.state.edges if e.dst == node.id
        ]
        if wires:
            lines.append("  wires: " + " · ".join(wires))
        return "\n".join(lines)

    def _node_name(self, node_id: str) -> str:
        node = self.state.nodes.get(node_id)
        return node.label if node is not None else node_id

    # ---- questions on the table ----

    def pending_ask(self) -> Ask | None:
        """The earliest open question that has rows to pick from.

        A question without options is asked in prose and answered in the
        message box; only one with them is a picker, and only one picker is
        ever on the table."""
        for q in self.state.questions:
            if q.open and q.options:
                return q.as_ask()
        return None

    def answer_ask(self, payload: dict[str, Any]) -> dict[str, Any]:
        """A row taken in the page's picker.

        Duck-typed like `apply_mutation`: the pump finds this by name on
        `ctx.arch`. It settles the question in the state and hands back the
        turn to send — the architect hears the answer as a pick, with the
        question it answers hung under it so it cannot bind it to the wrong
        one."""
        qid = str(payload.get("id") or "")
        q = self.state.question_by_id(qid)
        if q is None:
            known = ", ".join(x.id for x in self.state.questions) or "none"
            raise ValueError(f"no question {qid!r} (known: {known}).")
        ask = q.as_ask()
        picked = ask.take(str(payload.get("value") or ""))
        q.answer = picked.label
        q.status = "answered"
        self._record_pick(q, picked)
        self.touched("question", q.id)
        return {
            "answered": q.id,
            "value": picked.value,
            "label": picked.label,
            "input": f"{PICK_PREFIX}\n- {picked.label}\n  in answer to {q.id}: {q.question}",
        }

    def _record_pick(self, q: Any, picked: Any) -> None:
        """A row taken is a decision made. Record it here, deterministically,
        rather than hoping the architect calls `decide` afterwards — in the
        one live run before this existed, it did not, and the bundle would
        have shipped without the only call the user had actually made.

        The other rows travel as what it was weighed against, each with its
        one-line consequence. A revised pick updates the decision it made
        rather than adding a second one."""
        from .state import Decision, Option

        topic = q.summary or q.question
        options = [
            Option(name=o.label, pros=[o.description] if o.description else [])
            for o in q.options
        ]
        rationale = "picked by the user" + (f": {picked.description}" if picked.description else "")
        for d in self.state.decisions:
            if d.source == "user" and d.topic == topic:
                d.choice, d.options, d.rationale = picked.label, options, rationale
                return
        self.state.decisions.append(Decision(
            id=self.next_decision_id(), topic=topic, options=options,
            choice=picked.label, rationale=rationale, source="user",
        ))

    def next_question_id(self) -> str:
        return self.state.next_id("q", self.state.questions)

    def next_decision_id(self) -> str:
        return self.state.next_id("d", self.state.decisions)

    @classmethod
    def load(cls, run_dir: Path, **kw: Any) -> "ArchSession":
        """Restore persisted state for --resume; fresh state if none saved.
        A state file from the pre-rebuild harness raises LegacyStateError — see
        `ArchState.from_dict`."""
        path = run_dir / STATE_FILENAME
        state = None
        if path.is_file():
            state = ArchState.from_dict(json.loads(path.read_text(encoding="utf-8")))
        return cls(state=state, run_dir=run_dir, **kw)
